### PR TITLE
New branch

### PR DESCRIPTION
**Pull request by AlePuf**

### CODE REVIEW

**Comment by malinaslevoaca1** [comment]
is there a reason why you add the SPA files?

File: src/server.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import {rovers} from './rovers';
+import {photos} from './photos';
+
+const app = express();
+const port = 8000;
+const api_key = 'M3d4mWdNB9BtMBlncFSJ6BBy9hdlGmpOll5BJlTf';
+const api_base = 'https://api.nasa.gov/mars-photos/api/v1';
+const rover = 'curiosity';
+const camera_type = 'NAVCAM';
+
```
Comment (malinaslevoaca1) [inline src/server.ts]: are these 2 used anywhere?

File: src/server.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+const router = express.Router();
+router.get('/test', (req, res) => res.send('Hello world !'));
+app.use('/', router);
+
+router.get('/rovers', (req, res) => {
+    let api_url = `${api_base}/rovers?api_key=${api_key}`;
+    rovers(api_url, res);
+});
+
+router.get('/rovers/:rover/photos/:sol', (req, res) => {
```
Comment (malinaslevoaca1) [inline src/server.ts]: what happens if the rover is invalid?

File: src/photos.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        return;
+    }
+    if (page) {
+        api_url += `&page=${page}`;
+    }
+    if (camera) {
+        api_url += `&camera=${req.query.camera}`;
+    }
+    if (paginationStart && paginationEnd) {
+        if (typeof paginationStart == "string" && typeof paginationEnd == "string") {
```
Comment (malinaslevoaca1) [inline src/photos.ts]: why do we need this?

File: src/rovers.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    axios.get(api_url).then(resp => {
+        let data = resp.data.rovers;
+        let response = '';
+        data.forEach(function (r: JSONRovers) {
+            response += `ID: ${r.id}<br>Name: ${r.name}<br>Landing date: ${r.landing_date}<br>Launch date: ${r.launch_date}<br>Status: ${r.status}<br>Max sol: ${r.max_sol}<br>Max date: ${r.max_date}<br>Total photos: ${r.total_photos}<br>Cameras:<br>`;
+            r.cameras.forEach(function(c: JSONCamera) {
+                response += `&emsp;ID: ${c.id}<br>&emsp;Name: ${c.name}<br>&emsp;Full name: ${c.full_name}<br><br>`;
+            });
+        });
+        res.send(response);
```
Comment (malinaslevoaca1) [inline src/rovers.ts]: we should return a JSON instead

**Review by malinaslevoaca1 — COMMENTED** [review]
take a look at https://medium.com/codechef-vit/a-better-project-structure-with-express-and-node-js-c23abc2d736f - can you please refactor the project to have a similar structure?

**Comment by AlePuf** [comment]
> is there a reason why you add the SPA files?

nope it was by mistake, I accidentally saved the SPA project in this folder, will delete them now

File: src/photos.ts (hunk 3 above)
Comment (AlePuf) [reply src/photos.ts]: the parseInt function used for going through the pages throws a type error, so I had to add that constraint

Comment (malinaslevoaca1) [reply src/photos.ts]: I see, you don't have an else block, though - could you combine the 2 conditions (this and the one above)?

File: src/photos.ts [hunk 5]
```diff
@@ -22,34 +22,26 @@ export function photos(api_url: string, req: any, res: any) {
...
-            let data = resp.data.photos;
-            let response = '';
-            data.forEach(function (p: JSONPhoto) {
-                response += `Photo ID: ${p.id}<br>Camera: ${p.camera.name}<br><img src="${p.img_src}"><br><br>`;
-            });
-            res.send(response);
+            res.send(resp.data.photos);
         }).catch(() => {
-            console.log("error");
+            res.send("Error: invalid parameters");
```
Comment (malinaslevoaca1) [inline src/photos.ts]: are all the errors caught here thrown due to some invalid parameters?

File: .gitignore [hunk 6]
```diff
@@ -40,6 +40,7 @@ build/Release
 # Dependency directories
 node_modules/
 jspm_packages/
+marsappspa/
```
Comment (malinaslevoaca1) [inline .gitignore:43]: I think you want to delete it

File: src/Middleware/check_valid_rover.middleware.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import {API_KEY, API_BASE} from '../Routes/server';
+import {JSONRovers} from '../Models/interfaces';
+import axios from 'axios';
+
+export async function check_valid_rover(rover: string) : Promise<Boolean> {
```
Comment (malinaslevoaca1) [inline src/Middleware/check_valid_rover.middleware.ts:5]: this is not a middleware, is it?

check https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/Introduction#using_middleware

File: src/Middleware/check_valid_rover.middleware.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    let api_url = `${API_BASE}/rovers?api_key=${API_KEY}`;
+    let returnValue = false;
+    await axios.get(api_url).then(resp => {
+       let rovers : Array<JSONRovers> = resp.data.rovers;
+       for (let i = 0; i < rovers.length; i++) {
```
Comment (malinaslevoaca1) [inline src/Middleware/check_valid_rover.middleware.ts:10]: take a look at [some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

File: src/Models/interfaces.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+export interface JSONCamera {
```
Comment (malinaslevoaca1) [inline src/Models/interfaces.ts:1]: ideally (almost) each interface is in its own separate file

File: src/Controllers/rovers.controllers.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import axios from 'axios';
+import {JSONRovers, JSONCamera} from '../Models/interfaces';
+import {API_BASE, API_KEY} from '../Routes/server';
+
+export function rovers(req: any, res: any) {
+    let api_url = `${API_BASE}/rovers?api_key=${API_KEY}`;
+    axios.get(api_url).then(resp => {
```
Comment (malinaslevoaca1) [inline src/Controllers/rovers.controllers.ts:7]: you could add a service used for fetching data from the NASA API

File: src/Controllers/photos.controllers.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import axios from 'axios';
+import {JSONPhoto} from '../Models/interfaces';
+import {check_valid_rover} from '../Middleware/check_valid_rover.middleware';
+import {API_BASE, API_KEY} from '../Routes/server'
+
+export async function photos(req: any, res: any) {
```
Comment (malinaslevoaca1) [inline src/Controllers/photos.controllers.ts:6]: this function is a little long. Some suggestions:
- move validation to a separate file
- fetching data: use a service

File: src/Controllers/photos.controllers.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    if (paginationStart && paginationEnd && typeof paginationStart == "string" && typeof paginationEnd == "string") {
+        let response: Array<JSONPhoto> = [];
+        let promiseArray: any[] = [];
+        for (let i = parseInt(paginationStart); i <= parseInt(paginationEnd); i++) {
+            let url = api_url + `&page=${i}`;
+            promiseArray.push(axios.get(url));
+        }
+        Promise.all(promiseArray).then(resp => {
+            for (let i = 0; i < promiseArray.length; i++) {
```
Comment (malinaslevoaca1) [inline src/Controllers/photos.controllers.ts:39]: could use `map`

File: src/Controllers/photos.controllers.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        res.send("Error: must have both paginationStart and paginationEnd!");
+        return;
+    }
+    if (page) {
+        api_url += `&page=${page}`;
+    }
+    if (camera) {
+        api_url += `&camera=${req.query.camera}`;
+    }
+    if (paginationStart && paginationEnd && typeof paginationStart == "string" && typeof paginationEnd == "string") {
```
Comment (malinaslevoaca1) [inline src/Controllers/photos.controllers.ts:31]: is this still needed `paginationStart && paginationEnd`?

File: src/Routes/server.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+app.use(express.json());
+const router = express.Router();
+app.use('/', router);
+
+router.get('/test', (req, res) => res.send('Hello world !'));
+
+router.get('/rovers', rovers);
+
+router.get('/rovers/:rover/photos/:sol', photos);
```
Comment (malinaslevoaca1) [inline src/Routes/server.ts:18]: the routes could be moved to a separate file but configuring the server should still be in the main file (`server.ts` in our case)

File: src/Controllers/rovers.controllers.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import axios from 'axios';
+import {JSONRovers, JSONCamera} from '../Models/interfaces';
+import {API_BASE, API_KEY} from '../Routes/server';
+
+export function rovers(req: any, res: any) {
```
Comment (malinaslevoaca1) [inline src/Controllers/rovers.controllers.ts:5]: the function name doesn't describe what the function does - `getRovers` will be better, what do you think?

**Review by malinaslevoaca1 — APPROVED** [review]
Some suggestions - no need to make the changes unless you run out of work

**Pull request merged**